In [2]:
import numpy as np
import pandas as pd

In [145]:
df = pd.read_csv("loan_data_set.csv")

In [43]:
# print(df.head)

<bound method NDFrame.head of       Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849            

In [4]:
# df.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

In [146]:
Y = df['Loan_Status']
X = df.drop(columns =['Loan_ID','Loan_Status'])
print(X.columns)

Index(['Gender', 'Married', 'Dependents', 'Education', 'Self_Employed',
       'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area'],
      dtype='object')


In [6]:
#X.describe()
#df['Gender'].unique()

In [7]:
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(rows['Gender']!='Male' and rows['Gender']!='Female'):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.021172638436482084
[23, 126, 171, 188, 314, 334, 460, 467, 477, 507, 576, 588, 592]


In [8]:
# df['Married'].unique()

array(['No', 'Yes', nan], dtype=object)

In [9]:
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(rows['Married']!='Yes' and rows['Married']!='No'):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.004885993485342019
[104, 228, 435]


In [10]:
# df['Dependents'].unique()

array(['0', '1', '2', '3+', nan], dtype=object)

In [11]:
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(rows['Dependents'] not in ['0','1','2','3+']):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.024429967426710098
[102, 104, 120, 226, 228, 293, 301, 332, 335, 346, 355, 435, 517, 571, 597]


In [12]:
# df['Education'].unique()

array(['Graduate', 'Not Graduate'], dtype=object)

In [13]:
# df['Self_Employed'].unique()

array(['No', 'Yes', nan], dtype=object)

In [14]:
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(rows['Self_Employed'] not in ['Yes','No']):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.05211726384364821
[11, 19, 24, 29, 30, 95, 107, 111, 114, 158, 170, 218, 231, 236, 268, 295, 333, 336, 344, 374, 380, 385, 411, 432, 447, 463, 468, 535, 542, 579, 600, 601]


In [15]:
# import math
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(math.isnan(rows['ApplicantIncome'])):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.0
[]


In [16]:
# import math
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(math.isnan(rows['CoapplicantIncome'])):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.0
[]


In [17]:
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(math.isnan(rows['LoanAmount'])):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.035830618892508145
[0, 35, 63, 81, 95, 102, 103, 113, 127, 202, 284, 305, 322, 338, 387, 435, 437, 479, 524, 550, 551, 605]


In [18]:
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(math.isnan(rows['Loan_Amount_Term'])):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.02280130293159609
[19, 36, 44, 45, 73, 112, 165, 197, 223, 232, 335, 367, 421, 423]


In [19]:
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(math.isnan(rows['Credit_History'])):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.08143322475570032
[16, 24, 30, 42, 79, 83, 86, 95, 117, 125, 129, 130, 156, 181, 187, 198, 219, 236, 237, 259, 260, 279, 309, 313, 317, 318, 323, 348, 363, 377, 392, 395, 411, 444, 449, 451, 460, 473, 490, 491, 497, 503, 506, 530, 533, 544, 556, 565, 583, 600]


In [20]:
# df['Property_Area'].unique()

array(['Urban', 'Rural', 'Semiurban'], dtype=object)

In [21]:
# total=0
# count=0
# indices=[]
# for index,rows in df.iterrows():
#     if(rows['Property_Area'] not in ['Urban','Rural','Semiurban']):
#         count+=1
#         indices.append(index)
#     total+=1
# print(count/total)
# print(indices)

0.0
[]


### Exploratory Data Analysis
<table>
    <tr>
        <th>Column</th>
        <th>Unique values</th>
        <th>% of NaN</th>
        <th>Indices of NaN</th>
    </tr>
    <tr>
        <td>Gender</td>
        <td>Male,Female</td>
        <td>0.021</td>
        <td>[23, 126, 171, 188, 314, 334, 460, 467, 477, 507, 576, 588, 592]</td>
    </tr>
    <tr>
        <td>Married</td>
        <td>Yes,No</td>
        <td>0.0048</td>
        <td>[104, 228, 435]</td>
    </tr>
    <tr>
        <td>Dependents</td>
        <td>0,1,2,3+</td>
        <td>0.024</td>
        <td>[102, 104, 120, 226, 228, 293, 301, 332, 335, 346, 355, 435, 517, 571, 597]</td>
    </tr>
    <tr>
        <td>Education</td>
        <td>Graduate, Not Graduate</td>
        <td>0.0</td>
        <td>[]</td>
    </tr>
    <tr>
        <td>Self_Employed</td>
        <td>Yes, No</td>
        <td>0.052</td>
        <td>[11, 19, 24, 29, 30, 95, 107, 111, 114, 158, 170, 218, 231, 236, 268, 295, 333, 336, 344, 374, 380, 385, 411, 432, 447, 463, 468, 535, 542, 579, 600, 601]</td>
    </tr>
    <tr>
        <td>Applicant_Income</td>
        <td>some range</td>
        <td>0.0</td>
        <td>[]</td>
    </tr>
    <tr>
        <td>Coapplicant_Income</td>
        <td>some range</td>
        <td>0.0</td>
        <td>[]</td>
    </tr>
    <tr>
        <td>Loan_Amount</td>
        <td>some range</td>
        <td>0.036</td>
        <td>[0, 35, 63, 81, 95, 102, 103, 113, 127, 202, 284, 305, 322, 338, 387, 435, 437, 479, 524, 550, 551, 605]</td>
    </tr>
    <tr>
        <td>Loan_Amount_Term</td>
        <td>some range</td>
        <td>0.023</td>
        <td>[19, 36, 44, 45, 73, 112, 165, 197, 223, 232, 335, 367, 421, 423]</td>
    </tr>
    <tr>
        <td>Credit_History</td>
        <td>some range</td>
        <td>0.081</td>
        <td>[16, 24, 30, 42, 79, 83, 86, 95, 117, 125, 129, 130, 156, 181, 187, 198, 219, 236, 237, 259, 260, 279, 309, 313, 317, 318, 323, 348, 363, 377, 392, 395, 411, 444, 449, 451, 460, 473, 490, 491, 497, 503, 506, 530, 533, 544, 556, 565, 583, 600]</td>
    </tr>
    <tr>
        <td>Property_Area</td>
        <td>Urban,Rural,SemiUrban</td>
        <td>0.0</td>
        <td>[]</td>
    </tr>
</table>

In [22]:
# df['Loan_Amount_Term'].unique()

array([360., 120., 240.,  nan, 180.,  60., 300., 480.,  36.,  84.,  12.])

In [23]:
# len(df)

614

In [147]:
df['Gender'] = df['Gender'].map({'Female': 1, 'Male': 0})
df['Married'] = df['Married'].map({'Yes': 1, 'No': 0})
df['Dependents'] = df['Dependents'].map({'0':0,'1': 1, '2': 2,'3+':3})
df['Self_Employed'] = df['Self_Employed'].map({'Yes': 1, 'No': 0})
df['Property_Area'] = df['Property_Area'].map({'Urban':1,'Rural':2,'Semiurban':3})
df['Education'] = df['Education'].map({'Graduate': 1, 'Not Graduate': 0})
df['Loan_Status'] = df['Loan_Status'].map({'Y': 1, 'N': 0})

In [148]:
print(df.head)

<bound method NDFrame.head of       Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
0    LP001002     0.0      0.0         0.0          1            0.0   
1    LP001003     0.0      1.0         1.0          1            0.0   
2    LP001005     0.0      1.0         0.0          1            1.0   
3    LP001006     0.0      1.0         0.0          0            0.0   
4    LP001008     0.0      0.0         0.0          1            0.0   
..        ...     ...      ...         ...        ...            ...   
609  LP002978     1.0      0.0         0.0          1            0.0   
610  LP002979     0.0      1.0         3.0          1            0.0   
611  LP002983     0.0      1.0         1.0          1            0.0   
612  LP002984     0.0      1.0         2.0          1            0.0   
613  LP002990     1.0      0.0         0.0          1            1.0   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849            

In [46]:
# indices = [] 
# cols = df.columns
# cols=cols[1:]
# for index, row in df.iterrows():
#     flag=0
#     for col in cols:
# #         print(row[col],col)
#         if(math.isnan(row[col])):
#             flag=1
#             #print(row[col])
#     if(flag==1):
#         indices.append(index)
# print(len(indices))
# print(indices)

134
[0, 11, 16, 19, 23, 24, 29, 30, 35, 36, 42, 44, 45, 63, 73, 79, 81, 83, 86, 95, 102, 103, 104, 107, 111, 112, 113, 114, 117, 120, 125, 126, 127, 129, 130, 156, 158, 165, 170, 171, 181, 187, 188, 197, 198, 202, 218, 219, 223, 226, 228, 231, 232, 236, 237, 259, 260, 268, 279, 284, 293, 295, 301, 305, 309, 313, 314, 317, 318, 322, 323, 332, 333, 334, 335, 336, 338, 344, 346, 348, 355, 363, 367, 374, 377, 380, 385, 387, 392, 395, 411, 421, 423, 432, 435, 437, 444, 447, 449, 451, 460, 463, 467, 468, 473, 477, 479, 490, 491, 497, 503, 506, 507, 517, 524, 530, 533, 535, 542, 544, 550, 551, 556, 565, 571, 576, 579, 583, 588, 592, 597, 600, 601, 605]


In [47]:
# print(df.head)

<bound method NDFrame.head of       Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
0    LP001002     0.0      0.0         0.0          1            0.0   
1    LP001003     0.0      1.0         1.0          1            0.0   
2    LP001005     0.0      1.0         0.0          1            1.0   
3    LP001006     0.0      1.0         0.0          0            0.0   
4    LP001008     0.0      0.0         0.0          1            0.0   
..        ...     ...      ...         ...        ...            ...   
609  LP002978     1.0      0.0         0.0          1            0.0   
610  LP002979     0.0      1.0         3.0          1            0.0   
611  LP002983     0.0      1.0         1.0          1            0.0   
612  LP002984     0.0      1.0         2.0          1            0.0   
613  LP002990     1.0      0.0         0.0          1            1.0   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849            

In [149]:
# df = df.fillna(0)
df = df.dropna()

In [49]:
# df.drop(['Loan_ID'],axis=1)

Gender  Married  Dependents  Education  Self_Employed  ApplicantIncome  \
0       0.0      0.0         0.0          1            0.0             5849   
1       0.0      1.0         1.0          1            0.0             4583   
2       0.0      1.0         0.0          1            1.0             3000   
3       0.0      1.0         0.0          0            0.0             2583   
4       0.0      0.0         0.0          1            0.0             6000   
..      ...      ...         ...        ...            ...              ...   
609     1.0      0.0         0.0          1            0.0             2900   
610     0.0      1.0         3.0          1            0.0             4106   
611     0.0      1.0         1.0          1            0.0             8072   
612     0.0      1.0         2.0          1            0.0             7583   
613     1.0      0.0         0.0          1            1.0             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                  0.0         NaN             360.0             1.0   
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

     Property_Area  Loan_Status  
0                1            1  
1                2            0  
2                1            1  
3                1            1  
4                1            1  
..             ...          ...  
609              2            1  
610              2            1  
611              1            1  
612              1            1  
613              3            0  

[614 rows x 12 columns]

In [150]:
Y = df['Loan_Status']
X = df.drop(['Loan_Status','Loan_ID'],axis=1)

In [103]:
# Y

0      1
1      0
2      1
3      1
4      1
      ..
609    1
610    1
611    1
612    1
613    0
Name: Loan_Status, Length: 614, dtype: int64

In [59]:
# X

Gender  Married  Dependents  Education  Self_Employed  ApplicantIncome  \
1       0.0      1.0         1.0          1            0.0             4583   
2       0.0      1.0         0.0          1            1.0             3000   
3       0.0      1.0         0.0          0            0.0             2583   
4       0.0      0.0         0.0          1            0.0             6000   
5       0.0      1.0         2.0          1            1.0             5417   
..      ...      ...         ...        ...            ...              ...   
609     1.0      0.0         0.0          1            0.0             2900   
610     0.0      1.0         3.0          1            0.0             4106   
611     0.0      1.0         1.0          1            0.0             8072   
612     0.0      1.0         2.0          1            0.0             7583   
613     1.0      0.0         0.0          1            1.0             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
5               4196.0       267.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

     Property_Area  
1                2  
2                1  
3                1  
4                1  
5                1  
..             ...  
609              2  
610              2  
611              1  
612              1  
613              3  

[480 rows x 11 columns]

In [151]:
from sklearn.model_selection import train_test_split
x, x_unseen, y, y_unseen = train_test_split(X, Y, test_size=0.05, random_state=42)

In [63]:
# x_unseen

Gender  Married  Dependents  Education  Self_Employed  ApplicantIncome  \
92      0.0      1.0         2.0          0            0.0             3273   
529     0.0      0.0         0.0          0            0.0             6783   
505     0.0      1.0         2.0          1            0.0             3510   
358     0.0      1.0         0.0          0            0.0             3000   
512     0.0      1.0         2.0          1            0.0             3283   
220     0.0      1.0         0.0          1            0.0             2221   
347     0.0      1.0         2.0          0            0.0             3083   
482     0.0      1.0         0.0          1            0.0             2083   
97      0.0      1.0         0.0          1            0.0             1977   
566     0.0      0.0         0.0          1            0.0             3333   
10      0.0      1.0         2.0          1            0.0             3200   
362     0.0      1.0         0.0          1            0.0             4750   
555     0.0      1.0         1.0          1            0.0             5468   
384     0.0      1.0         2.0          1            0.0             4566   
40      0.0      0.0         0.0          1            0.0             3600   
123     0.0      1.0         2.0          1            0.0             2957   
69      1.0      0.0         0.0          1            0.0             4300   
541     1.0      0.0         0.0          1            0.0             2138   
46      0.0      1.0         1.0          1            0.0             5649   
258     0.0      1.0         0.0          1            0.0            14683   
233     1.0      0.0         0.0          1            0.0             8333   
595     0.0      0.0         0.0          0            0.0             3833   
89      0.0      1.0         0.0          1            0.0             5695   
337     0.0      1.0         2.0          1            1.0             2500   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
92              1820.0        81.0             360.0             1.0   
529                0.0       130.0             360.0             1.0   
505             4416.0       243.0             360.0             1.0   
358             1666.0       100.0             480.0             0.0   
512             2035.0       148.0             360.0             1.0   
220                0.0        60.0             360.0             0.0   
347             2168.0       126.0             360.0             1.0   
482             3150.0       128.0             360.0             1.0   
97               997.0        50.0             360.0             1.0   
566                0.0        70.0             360.0             1.0   
10               700.0        70.0             360.0             1.0   
362             2333.0       130.0             360.0             1.0   
555             1032.0        26.0             360.0             1.0   
384                0.0       100.0             360.0             1.0   
40                 0.0        80.0             360.0             1.0   
123                0.0        81.0             360.0             1.0   
69                 0.0       136.0             360.0             0.0   
541                0.0        99.0             360.0             0.0   
46                 0.0        44.0             360.0             1.0   
258             2100.0       304.0             360.0             1.0   
233                0.0       280.0             360.0             1.0   
595                0.0       110.0             360.0             1.0   
89              4167.0       175.0             360.0             1.0   
337             4600.0       176.0             360.0             1.0   

     Property_Area  
92               1  
529              3  
505              2  
358              1  
512              1  
220              1  
347              1  
482              3  
97               3  
566              1

In [152]:
X_train, X_test, Y_train, Y_test = train_test_split(x, y, test_size=0.33, random_state=42)

In [68]:
# Y_train

489    0
483    1
263    1
569    0
379    1
      ..
124    1
252    1
80     1
192    0
161    0
Name: Loan_Status, Length: 305, dtype: int64

In [114]:
import numpy as np
import matplotlib.pyplot as plt
from keras.models import Sequential
from keras.layers import Dense
from keras import optimizers

# KERAS STARTS HERE

# Create some data for training
# X_train = np.linspace(0,10,200)
# Y_train = np.cos(X_train) + np.random.randn(200)*0.15
# +np.random.randn adds noise

#plt.scatter(X_train,Y_train,color="red")
#plt.show()

# Create a model for training
model = Sequential()
# just give me the layers, I'll connect it

# FIRST HIDDEN LAYER
model.add(Dense(10,input_dim=11,activation = 'tanh'))  
# add - adds a layer
# dense - fully connected
# 10 - How many neurons in this
# input dimensions and activation function

# SECOND HIDDEN LAYER
model.add(Dense(20,activation='tanh'))
# don't need to specify input dimensions - it will take from previous layer

# THIRD HIDDEN LAYER
#model.add(Dense(20,activation = 'tanh'))

#OUTPUT LAYER
model.add(Dense(1,activation='sigmoid'))

model.summary()
# None refers to the number of instances
# 1st layer: 10+10
# 2nd layer: 10*20+20
# 3rd layer: 20*20+20
# 4th layer: 20*1+1

# Compile the model
opt = optimizers.SGD(momentum=0)
model.compile(optimizer=opt,loss='mse')

# or

#model.compile(optimizer='sgd',loss='mse')

# Train now
model.fit(X_train,Y_train,epochs = 1500)

# Predict y values
# Checking how model learnt
#Y_pred = model.predict(X_train)

#plt.scatter(X_train,Y_train,color="red")
#plt.plot(X_train,Y_pred,color="blue")
#plt.show()

# Check how it works on unseen data
# unseen data must be in same range - train on b/w images of cats and 
# test on color images will not work well obviously
# X_test=np.linspace(0,10,100)
Y_pred = model.predict(X_test)
#plt.scatter(X_train,Y_train,color="red")
# plt.scatter(X_test,Y_pred,color="blue")
#plt.show()


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_40 (Dense)             (None, 10)                120       
_________________________________________________________________
dense_41 (Dense)             (None, 20)                220       
_________________________________________________________________
dense_42 (Dense)             (None, 1)                 21        
Total params: 361
Trainable params: 361
Non-trainable params: 0
_________________________________________________________________
Epoch 1/1500
390/390 [==============================] - 1s 3ms/step - loss: 0.2947
Epoch 2/1500
390/390 [==============================] - 0s 94us/step - loss: 0.2774
Epoch 3/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2655
Epoch 4/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2587
Epoch 5/1500
390/390 [==============================] - 0s 80us/step - lo

390/390 [==============================] - 0s 83us/step - loss: 0.2192
Epoch 89/1500
390/390 [==============================] - 0s 91us/step - loss: 0.2192
Epoch 90/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2192
Epoch 91/1500
390/390 [==============================] - 0s 78us/step - loss: 0.2191
Epoch 92/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2192
Epoch 93/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2192
Epoch 94/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2192
Epoch 95/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2192
Epoch 96/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2191
Epoch 97/1500
390/390 [==============================] - 0s 73us/step - loss: 0.2191
Epoch 98/1500
390/390 [==============================] - 0s 75us/step - loss: 0.2191
Epoch 99/1500
390/390 [==============================] - 0s 74us/step - loss: 0

390/390 [==============================] - 0s 81us/step - loss: 0.2186
Epoch 184/1500
390/390 [==============================] - 0s 96us/step - loss: 0.2186
Epoch 185/1500
390/390 [==============================] - 0s 86us/step - loss: 0.2187
Epoch 186/1500
390/390 [==============================] - 0s 93us/step - loss: 0.2186
Epoch 187/1500
390/390 [==============================] - 0s 89us/step - loss: 0.2186
Epoch 188/1500
390/390 [==============================] - 0s 90us/step - loss: 0.2187
Epoch 189/1500
390/390 [==============================] - 0s 87us/step - loss: 0.2186
Epoch 190/1500
390/390 [==============================] - 0s 91us/step - loss: 0.2186
Epoch 191/1500
390/390 [==============================] - 0s 87us/step - loss: 0.2186
Epoch 192/1500
390/390 [==============================] - 0s 78us/step - loss: 0.2186
Epoch 193/1500
390/390 [==============================] - 0s 77us/step - loss: 0.2186
Epoch 194/1500
390/390 [==============================] - 0s 74us/ste

390/390 [==============================] - 0s 153us/step - loss: 0.2182
Epoch 279/1500
390/390 [==============================] - 0s 131us/step - loss: 0.2182
Epoch 280/1500
390/390 [==============================] - 0s 114us/step - loss: 0.2182
Epoch 281/1500
390/390 [==============================] - 0s 110us/step - loss: 0.2182
Epoch 282/1500
390/390 [==============================] - 0s 110us/step - loss: 0.2182
Epoch 283/1500
390/390 [==============================] - 0s 92us/step - loss: 0.2182
Epoch 284/1500
390/390 [==============================] - 0s 94us/step - loss: 0.2182
Epoch 285/1500
390/390 [==============================] - 0s 132us/step - loss: 0.2182
Epoch 286/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2182
Epoch 287/1500
390/390 [==============================] - 0s 113us/step - loss: 0.2182
Epoch 288/1500
390/390 [==============================] - 0s 103us/step - loss: 0.2182
Epoch 289/1500
390/390 [==============================] - 0s 

390/390 [==============================] - 0s 84us/step - loss: 0.2227
Epoch 374/1500
390/390 [==============================] - 0s 80us/step - loss: 0.2228
Epoch 375/1500
390/390 [==============================] - 0s 88us/step - loss: 0.2226
Epoch 376/1500
390/390 [==============================] - 0s 102us/step - loss: 0.2226
Epoch 377/1500
390/390 [==============================] - 0s 105us/step - loss: 0.2226
Epoch 378/1500
390/390 [==============================] - 0s 99us/step - loss: 0.2225
Epoch 379/1500
390/390 [==============================] - 0s 95us/step - loss: 0.2225
Epoch 380/1500
390/390 [==============================] - 0s 95us/step - loss: 0.2224
Epoch 381/1500
390/390 [==============================] - 0s 92us/step - loss: 0.2224
Epoch 382/1500
390/390 [==============================] - 0s 94us/step - loss: 0.2224
Epoch 383/1500
390/390 [==============================] - 0s 90us/step - loss: 0.2223
Epoch 384/1500
390/390 [==============================] - 0s 87us/s

Epoch 468/1500
390/390 [==============================] - 0s 100us/step - loss: 0.2209
Epoch 469/1500
390/390 [==============================] - 0s 96us/step - loss: 0.2209
Epoch 470/1500
390/390 [==============================] - 0s 98us/step - loss: 0.2209
Epoch 471/1500
390/390 [==============================] - 0s 92us/step - loss: 0.2209
Epoch 472/1500
390/390 [==============================] - 0s 94us/step - loss: 0.2209
Epoch 473/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2208
Epoch 474/1500
390/390 [==============================] - 0s 88us/step - loss: 0.2208
Epoch 475/1500
390/390 [==============================] - 0s 95us/step - loss: 0.2208
Epoch 476/1500
390/390 [==============================] - 0s 97us/step - loss: 0.2208
Epoch 477/1500
390/390 [==============================] - 0s 101us/step - loss: 0.2208
Epoch 478/1500
390/390 [==============================] - 0s 89us/step - loss: 0.2208
Epoch 479/1500
390/390 [============================

Epoch 563/1500
390/390 [==============================] - 0s 88us/step - loss: 0.2202
Epoch 564/1500
390/390 [==============================] - 0s 85us/step - loss: 0.2202
Epoch 565/1500
390/390 [==============================] - 0s 86us/step - loss: 0.2202
Epoch 566/1500
390/390 [==============================] - 0s 86us/step - loss: 0.2202
Epoch 567/1500
390/390 [==============================] - 0s 87us/step - loss: 0.2203
Epoch 568/1500
390/390 [==============================] - 0s 85us/step - loss: 0.2202
Epoch 569/1500
390/390 [==============================] - 0s 80us/step - loss: 0.2202
Epoch 570/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2202
Epoch 571/1500
390/390 [==============================] - 0s 85us/step - loss: 0.2202
Epoch 572/1500
390/390 [==============================] - 0s 85us/step - loss: 0.2201
Epoch 573/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2202
Epoch 574/1500
390/390 [==============================

390/390 [==============================] - 0s 93us/step - loss: 0.2199
Epoch 659/1500
390/390 [==============================] - 0s 91us/step - loss: 0.2198
Epoch 660/1500
390/390 [==============================] - 0s 74us/step - loss: 0.2198
Epoch 661/1500
390/390 [==============================] - 0s 75us/step - loss: 0.2199
Epoch 662/1500
390/390 [==============================] - 0s 76us/step - loss: 0.2199
Epoch 663/1500
390/390 [==============================] - 0s 74us/step - loss: 0.2199
Epoch 664/1500
390/390 [==============================] - 0s 88us/step - loss: 0.2198
Epoch 665/1500
390/390 [==============================] - 0s 109us/step - loss: 0.2198
Epoch 666/1500
390/390 [==============================] - 0s 123us/step - loss: 0.2198
Epoch 667/1500
390/390 [==============================] - 0s 114us/step - loss: 0.2198
Epoch 668/1500
390/390 [==============================] - 0s 115us/step - loss: 0.2199
Epoch 669/1500
390/390 [==============================] - 0s 109u

390/390 [==============================] - 0s 98us/step - loss: 0.2197
Epoch 753/1500
390/390 [==============================] - 0s 87us/step - loss: 0.2196
Epoch 754/1500
390/390 [==============================] - 0s 77us/step - loss: 0.2197
Epoch 755/1500
390/390 [==============================] - 0s 75us/step - loss: 0.2197
Epoch 756/1500
390/390 [==============================] - 0s 74us/step - loss: 0.2197
Epoch 757/1500
390/390 [==============================] - 0s 73us/step - loss: 0.2196
Epoch 758/1500
390/390 [==============================] - 0s 73us/step - loss: 0.2197
Epoch 759/1500
390/390 [==============================] - 0s 77us/step - loss: 0.2197
Epoch 760/1500
390/390 [==============================] - 0s 75us/step - loss: 0.2197
Epoch 761/1500
390/390 [==============================] - 0s 74us/step - loss: 0.2196
Epoch 762/1500
390/390 [==============================] - 0s 73us/step - loss: 0.2197
Epoch 763/1500
390/390 [==============================] - 0s 79us/ste

390/390 [==============================] - 0s 82us/step - loss: 0.2193
Epoch 848/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2194
Epoch 849/1500
390/390 [==============================] - 0s 78us/step - loss: 0.2193
Epoch 850/1500
390/390 [==============================] - 0s 77us/step - loss: 0.2193
Epoch 851/1500
390/390 [==============================] - 0s 78us/step - loss: 0.2193
Epoch 852/1500
390/390 [==============================] - 0s 73us/step - loss: 0.2194
Epoch 853/1500
390/390 [==============================] - 0s 79us/step - loss: 0.2193
Epoch 854/1500
390/390 [==============================] - 0s 78us/step - loss: 0.2197
Epoch 855/1500
390/390 [==============================] - 0s 78us/step - loss: 0.2196
Epoch 856/1500
390/390 [==============================] - 0s 72us/step - loss: 0.2189
Epoch 857/1500
390/390 [==============================] - 0s 73us/step - loss: 0.2188
Epoch 858/1500
390/390 [==============================] - 0s 75us/ste

390/390 [==============================] - 0s 88us/step - loss: 0.2182
Epoch 943/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2183
Epoch 944/1500
390/390 [==============================] - 0s 77us/step - loss: 0.2182
Epoch 945/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2183
Epoch 946/1500
390/390 [==============================] - 0s 85us/step - loss: 0.2183
Epoch 947/1500
390/390 [==============================] - 0s 78us/step - loss: 0.2183
Epoch 948/1500
390/390 [==============================] - 0s 89us/step - loss: 0.2183
Epoch 949/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2183
Epoch 950/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2183
Epoch 951/1500
390/390 [==============================] - 0s 86us/step - loss: 0.2182
Epoch 952/1500
390/390 [==============================] - 0s 80us/step - loss: 0.2182
Epoch 953/1500
390/390 [==============================] - 0s 83us/ste

390/390 [==============================] - 0s 77us/step - loss: 0.2181
Epoch 1038/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2181
Epoch 1039/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2181
Epoch 1040/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2181
Epoch 1041/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2181
Epoch 1042/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2181
Epoch 1043/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2181
Epoch 1044/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2181
Epoch 1045/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2181
Epoch 1046/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2181
Epoch 1047/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2181
Epoch 1048/1500
390/390 [==============================] - 

390/390 [==============================] - 0s 80us/step - loss: 0.2180
Epoch 1132/1500
390/390 [==============================] - 0s 78us/step - loss: 0.2180
Epoch 1133/1500
390/390 [==============================] - 0s 80us/step - loss: 0.2180
Epoch 1134/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2180
Epoch 1135/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2180
Epoch 1136/1500
390/390 [==============================] - 0s 80us/step - loss: 0.2180
Epoch 1137/1500
390/390 [==============================] - 0s 78us/step - loss: 0.2180
Epoch 1138/1500
390/390 [==============================] - 0s 77us/step - loss: 0.2180
Epoch 1139/1500
390/390 [==============================] - 0s 77us/step - loss: 0.2180
Epoch 1140/1500
390/390 [==============================] - 0s 82us/step - loss: 0.2181
Epoch 1141/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2180
Epoch 1142/1500
390/390 [==============================] - 

Epoch 1225/1500
390/390 [==============================] - 0s 94us/step - loss: 0.2179
Epoch 1226/1500
390/390 [==============================] - 0s 79us/step - loss: 0.2179
Epoch 1227/1500
390/390 [==============================] - 0s 92us/step - loss: 0.2180
Epoch 1228/1500
390/390 [==============================] - 0s 96us/step - loss: 0.2180
Epoch 1229/1500
390/390 [==============================] - 0s 93us/step - loss: 0.2179
Epoch 1230/1500
390/390 [==============================] - 0s 103us/step - loss: 0.2179
Epoch 1231/1500
390/390 [==============================] - 0s 96us/step - loss: 0.2179
Epoch 1232/1500
390/390 [==============================] - 0s 100us/step - loss: 0.2180
Epoch 1233/1500
390/390 [==============================] - 0s 99us/step - loss: 0.2179
Epoch 1234/1500
390/390 [==============================] - 0s 102us/step - loss: 0.2179
Epoch 1235/1500
390/390 [==============================] - 0s 97us/step - loss: 0.2180
Epoch 1236/1500
390/390 [===============

390/390 [==============================] - 0s 111us/step - loss: 0.2179
Epoch 1319/1500
390/390 [==============================] - 0s 106us/step - loss: 0.2178
Epoch 1320/1500
390/390 [==============================] - 0s 95us/step - loss: 0.2178
Epoch 1321/1500
390/390 [==============================] - 0s 92us/step - loss: 0.2178
Epoch 1322/1500
390/390 [==============================] - 0s 92us/step - loss: 0.2179
Epoch 1323/1500
390/390 [==============================] - 0s 106us/step - loss: 0.2179
Epoch 1324/1500
390/390 [==============================] - 0s 102us/step - loss: 0.2178
Epoch 1325/1500
390/390 [==============================] - 0s 101us/step - loss: 0.2178
Epoch 1326/1500
390/390 [==============================] - 0s 95us/step - loss: 0.2179
Epoch 1327/1500
390/390 [==============================] - 0s 120us/step - loss: 0.2179
Epoch 1328/1500
390/390 [==============================] - 0s 107us/step - loss: 0.2178
Epoch 1329/1500
390/390 [===========================

390/390 [==============================] - 0s 112us/step - loss: 0.2178
Epoch 1412/1500
390/390 [==============================] - 0s 109us/step - loss: 0.2177
Epoch 1413/1500
390/390 [==============================] - 0s 97us/step - loss: 0.2178
Epoch 1414/1500
390/390 [==============================] - 0s 99us/step - loss: 0.2178
Epoch 1415/1500
390/390 [==============================] - 0s 91us/step - loss: 0.2178
Epoch 1416/1500
390/390 [==============================] - 0s 95us/step - loss: 0.2178
Epoch 1417/1500
390/390 [==============================] - 0s 100us/step - loss: 0.2178
Epoch 1418/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2178
Epoch 1419/1500
390/390 [==============================] - 0s 83us/step - loss: 0.2179
Epoch 1420/1500
390/390 [==============================] - 0s 81us/step - loss: 0.2179
Epoch 1421/1500
390/390 [==============================] - 0s 84us/step - loss: 0.2178
Epoch 1422/1500
390/390 [==============================]

In [115]:
# from sklearn.metrics import accuracy_score
# accuracy_score(Y_test, Y_pred)
# print(type(Y_pred))
Y= Y_test.as_matrix()
count=0
for i in range(len(Y_pred)):
#     print(Y_pred[i],Y[i])
    if(Y_pred[i]>0.5 and Y[i]==0):
        print(Y_pred[i])
        count+=1
print(count/len(Y))
# print(Y_pred)
# accuracy_score(Y, Y_pred)

[0.63813436]
[0.77114105]
[0.7600366]
[0.63813436]
[0.6275129]
[0.6404568]
[0.63813436]
[0.63813436]
[0.63813436]
[0.63813436]
[0.6404568]
[0.6404568]
[0.63813436]
[0.63813436]
[0.63813436]
[0.63813436]
[0.63813436]
[0.6404568]
[0.63813436]
[0.6404568]
[0.63813436]
[0.77114105]
[0.6404568]
[0.6404568]
[0.7600366]
[0.7600366]
[0.6404568]
[0.63813436]
[0.6404568]
[0.6404568]
[0.7600366]
[0.6404568]
[0.63813436]
[0.7600366]
[0.6404568]
[0.6404568]
[0.6404568]
[0.63813436]
[0.63813436]
[0.63813436]
[0.63813436]
[0.63813436]
[0.7600366]
[0.63813436]
[0.63813436]
[0.63813436]
[0.63813436]
[0.63813436]
[0.7600366]
[0.63813436]
[0.63813436]
[0.77114105]
[0.6404568]
[0.63813436]
[0.7600366]
[0.6275129]
[0.6404568]
0.29533678756476683
[[0.63813436]
 [0.7600366 ]
 [0.6404568 ]
 [0.77114105]
 [0.6404568 ]
 [0.7600366 ]
 [0.6404568 ]
 [0.6404568 ]
 [0.77114105]
 [0.7600366 ]
 [0.6404568 ]
 [0.6404568 ]
 [0.7600366 ]
 [0.6404568 ]
 [0.63813436]
 [0.63813436]
 [0.6404568 ]
 [0.63813436]
 [0.6404568 ]

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  after removing the cwd from sys.path.


In [128]:
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=51)
neigh.fit(X_train, Y_train)
X = X_test.values
count=0
for i in range(len(X_test)):
#     print(X_test)
    y = neigh.predict([X[i]])
#     print(y,Y[i])
    if(y[0]!=Y[i]):
        count+=1
print(count/len(X))

0.3005181347150259


In [153]:
from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor(n_estimators = 1000, random_state = 42)
rf.fit(X_train, Y_train)
pred = rf.predict(X_test)
errors = abs(pred - Y_test)

In [130]:
# mape = 100 * (errors / Y_test)
# accuracy = 100 - np.mean(mape)
# print('Accuracy:', round(accuracy, 2), '%.')

Accuracy: -inf %.


In [154]:
print(errors)

56     0.029
177    0.614
178    0.020
190    0.542
289    0.175
       ...  
445    0.038
308    0.189
310    0.185
466    0.369
589    0.007
Name: Loan_Status, Length: 151, dtype: float64


In [155]:
Y = Y_test.as_matrix()
fp=0
fn=0
threshold = 0.5
for i in range(len(Y)):
    if(Y[i]==1 and pred[i]<threshold):
#         print(pred[i],Y[i])
        fn+=1
    if(Y[i]==0 and pred[i]>threshold):
        fp+=1
#         print(pred[i],Y[i])
print(fn,fp)
print((fn+fp)/len(Y))

10 22
0.2119205298013245


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  """Entry point for launching an IPython kernel.


In [156]:
from sklearn.svm import SVC
svclassifier = SVC(kernel='linear')
svclassifier.fit(X_train, Y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='linear', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False)

In [ ]:
y_pred = svclassifier.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))